File: Modules/Relate.m
function handles = Relate(handles)

% Help for the Relate module:
% Category: Object Processing
%
% SHORT DESCRIPTION:
% Assigns relationships: All objects (e.g. speckles) within a parent object
% (e.g. nucleus) become its children.
% *************************************************************************
%
% Allows associating "children" objects with "parent" objects. This is
% useful for counting the number of children associated with each parent,
% and for calculating mean measurement values for all children that are
% associated with each parent. For every measurement that has been made of
% the children objects upstream in the pipeline, this module calculates the
% mean value of that measurement over all children and stores it as a
% measurement for the parent, as "Mean_<child>_<category>_<feature>". 
% For this reason, this module should be placed *after* all Measure modules
% that make measurements of the children objects.
%
% An object will be considered a child even if the edge is the only part
% touching a parent object. If an object is touching two parent objects,
% the objects parent will be the higher numbered parent.
%
% The minimum distances of each child to its parent are also calculated.
% These values are associated with the child objects. If an "Other" object
% is defined (e.g. Nuclei), then distances are calculated to this object
% too, as well as normalized distances.  Normalized distances for each
% child have a range [0 1] and are calculated as:
% (distance to the Parent) / sum(distances to parent and Other object)
%
% To access the Child/Parent label matrix image in downstream modules, use
% the "Other..." method to choose your image and type Parent_Child,
% where 'Parent' and 'Child' are the names of the objects as selected in 
% Relate's first two settings.  For example, if the parent objects are 
% "Cytoplasm" and the child objects are "Speckles", then downstream choose
% "Cytoplasm_Speckles".
%
% Measurement Categories (each with only one Feature):
% Parent, Children, SubObjectFlag, Distance, NormDistance

% CellProfiler is distributed under the GNU General Public License.
% See the accompanying file LICENSE for details.
%
% Developed by the Whitehead Institute for Biomedical Research.
% Copyright 2003,2004,2005.
%
% Please see the AUTHORS file for credits.
%
% Website: http://www.cellprofiler.org
%
% $Revision$

%%%%%%%%%%%%%%%%%
%%% VARIABLES %%%
%%%%%%%%%%%%%%%%%
drawnow

[CurrentModule, CurrentModuleNum, ModuleName] = CPwhichmodule(handles);

%textVAR01 = What objects are the children objects (subobjects)?
%infotypeVAR01 = objectgroup
%inputtypeVAR01 = popupmenu
SubObjectName = char(handles.Settings.VariableValues{CurrentModuleNum,1});

%textVAR02 = What are the parent objects?
%infotypeVAR02 = objectgroup
%inputtypeVAR02 = popupmenu
ParentName{1} = char(handles.Settings.VariableValues{CurrentModuleNum,2});

%textVAR03 = Do you want to find minimum distances of each child to its parent?
%choiceVAR03 = No
%choiceVAR03 = Yes
%inputtypeVAR03 = popupmenu
FindParentChildDistances = char(handles.Settings.VariableValues{CurrentModuleNum,3});

%textVAR04 = (If 'Yes' to above) What other object do you want to find distances to? (Must be one object per parent object, e.g. Nuclei)
%infotypeVAR04 = objectgroup
%choiceVAR04 = Do not use
%inputtypeVAR04 = popupmenu
ParentName{2} = char(handles.Settings.VariableValues{CurrentModuleNum,4});

%textVAR05 = Do you want to generate mean child measurements for all parents?
%choiceVAR05 = No
%choiceVAR05 = Yes
%inputtypeVAR05 = popupmenu
FindMeanMeasurements = char(handles.Settings.VariableValues{CurrentModuleNum,5});

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%VariableRevisionNumber = 3

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% PRELIMINARY CALCULATIONS %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
drawnow

% Do we want to calculate mean measurements?
wantMeanMeasurements = strncmpi(FindMeanMeasurements,'y',1);

% Do we want to calculate minimum distances?
wantMinDistances = strncmpi(FindParentChildDistances,'y',1);

%%% Retrieves the label matrix image that contains the edited primary
%%% segmented objects.
SubObjectLabelMatrix = CPretrieveimage(handles,['Segmented', SubObjectName],ModuleName,'MustBeGray','DontCheckScale');

%%% Retrieves the label matrix image that contains the edited primary
%%% segmented objects.
ParentObjectLabelMatrix = CPretrieveimage(handles,['Segmented', ParentName{1}],ModuleName,'MustBeGray','DontCheckScale');

%%% Retrieves the label matrix image that contains the edited primary
%%% segmented objects.
if ~strcmp(ParentName{2},'Do not use')

    % Sanity checks
    if strcmp(SubObjectName,ParentName{1}) || strcmp(SubObjectName,ParentName{2})
        CPwarndlg('The Children and at least one of the Parent objects are the same.  Your results may be erroneous.','Relate module')
    end
    if strcmp(ParentName{1},ParentName{2})
        CPwarndlg('The Parent and Other Object are the same.  Your results may be erroneous.','Relate module')
    end

    StepParentObjectLabelMatrix = CPretrieveimage(handles,['Segmented', ParentName{2}],ModuleName,'MustBeGray','DontCheckScale');

    % Sanity check
    if max(ParentObjectLabelMatrix(:)) ~= max(StepParentObjectLabelMatrix(:))
        CPwarndlg(['The number of Parent Objects (' num2str(max(ParentObjectLabelMatrix(:))) ...
            ') does not equal the number of Other objects (' num2str(max(StepParentObjectLabelMatrix(:))) ...
            ') in the Relate Module, Cycle#' num2str(handles.Current.SetBeingAnalyzed) ...
            '.  If the difference is large, this may cause the Relate Module output to be suspect.'],'Relate module')
    end
else
    ParentName = {ParentName{1}};
end

%%%%%%%%%%%%%%%%%%%%%%
%%% IMAGE ANALYSIS %%%
%%%%%%%%%%%%%%%%%%%%%%
drawnow

[handles,NumberOfChildren,ParentsOfChildren] = CPrelateobjects(handles,SubObjectName,ParentName{1},...
    SubObjectLabelMatrix,ParentObjectLabelMatrix,ModuleName);
handles = CPaddmeasurements(handles,SubObjectName,'SubObjectFlag',1);

if wantMinDistances
    % Save Distance 'Features'

    % Calcuate the smallest distance from each Child to their Parent
    % If no parent exists, then Distance = NaN

    if isfield(handles.Measurements.(SubObjectName),'Location_Center_X')

        for thisParent = ParentName %% Will need to change if we add more StepParents
            % Calculate perimeters for all parents simultaneously
            DistTransAll = CPlabelperim((CPretrieveimage(handles,['Segmented' thisParent{1}],ModuleName)));
            Dists = zeros(max(SubObjectLabelMatrix(:)), 1);
            if max(ParentsOfChildren) > 0,
                for iParentsOfChildren = 1:max(ParentsOfChildren)
                    % Calculate distance transform to perimeter of Parent objects
                    DistTrans = (bwdist(DistTransAll == iParentsOfChildren));

                    % Get location of each child object
                    ChList = find(ParentsOfChildren == iParentsOfChildren);
                    ChildrenLocationsX = handles.Measurements.(SubObjectName).Location_Center_X{handles.Current.SetBeingAnalyzed}(ChList,:);
                    ChildrenLocationsY = handles.Measurements.(SubObjectName).Location_Center_Y{handles.Current.SetBeingAnalyzed}(ChList,:);
                    roundedChLocX = round(ChildrenLocationsX);
                    roundedChLocY= round(ChildrenLocationsY);
                    idx = sub2ind(size(DistTrans),roundedChLocY(:,1), roundedChLocX(:,1));
                    Dist = DistTrans(idx);
                    Dists(ChList) = Dist;
                end
                handles = CPaddmeasurements(handles,SubObjectName,['Distance_' thisParent{1}], Dists);
            else
                handles = CPaddmeasurements(handles,SubObjectName,['Distance_' thisParent{1}], nan(max(length(NumberOfChildren),1), 1));
            end
        end
    else
        warning('There is no ''Location'' field with which to find subObj to Parent distances')
    end

    % Calculate normalized distances
    % All distances are relative to the *first* parent.
    if length(ParentName) > 1
        FirstParentDist =   handles.Measurements.(SubObjectName).(['Distance_',ParentName{1}]){handles.Current.SetBeingAnalyzed};
        OtherObjDist =      handles.Measurements.(SubObjectName).(['Distance_',ParentName{2}]){handles.Current.SetBeingAnalyzed};
        NormDist = FirstParentDist ./ sum([FirstParentDist OtherObjDist],2);
        NormDist(isnan(NormDist)) = 0;  %% In case sum(Dist,2) == 0 for any reason (no parents/child, or child touching either parent)

        % Save normalized distances
        handles = CPaddmeasurements(handles,SubObjectName, ['NormDistance_',ParentName{1}],NormDist);
    end
end

if wantMeanMeasurements
    % Adds a 'Mean_<SubObjectName>' field to the handles.Measurements structure
    % which finds the mean measurements of all the subObjects that relate to each parent object
    MeasurementFieldnames = fieldnames(handles.Measurements.(SubObjectName))';

    % Some measurments need to be excluded from the per-parent mean calculation
    ExcludedMeasurementsPrefixes = {'SubObjectFlag',...                                             % Child flags, which are a single scalar value (Relate)
        'Parent_','Children_',...                                                                   % Object lists and per-parent counts (CPRelateobjects)
        'Mean_',...                                                                                 % Per-parent mean measurments already calculated (Relate)
        'TrackObjects_Linearity_','TrackObjects_IntegratedDistance_','TrackObjects_Lifetime_'};     % Measurements which are calculated retrospectively (TrackObjects)
    
    if isfield(handles.Measurements.(SubObjectName),['Parent_',ParentName{1}])
        % Why is test line here? Isn't this always the case?  Or is it in case Relate is called twice?- Ray 2007-08-09
        if length(handles.Measurements.(SubObjectName).(CPjoinstrings('Parent_',ParentName{1}))) >= handles.Current.SetBeingAnalyzed
            Parents = handles.Measurements.(SubObjectName).(CPjoinstrings('Parent_',ParentName{1})){handles.Current.SetBeingAnalyzed};
            MeasurementFeatures = fieldnames(handles.Measurements.(SubObjectName));
            for i = 1:length(MeasurementFeatures)
                Fieldname = MeasurementFieldnames{i};
                if any(cell2mat(cellfun(@strncmp,   repmat({Fieldname},[1 length(ExcludedMeasurementsPrefixes)]),...
                                                    ExcludedMeasurementsPrefixes,...
                                                    num2cell(cellfun(@length,ExcludedMeasurementsPrefixes)),'UniformOutput',false)))
                    continue;
                end
                Measurements = handles.Measurements.(SubObjectName).(Fieldname){handles.Current.SetBeingAnalyzed};
                MeanVals = zeros(max(Parents), 1);
                if max(Parents) > 0
                    for j = 1:max(Parents),
                        indices = find(Parents == j);
                        if ~ isempty(indices),
                            MeanVals(j) = mean(Measurements(indices));
                        end
                    end
                end
                handles = CPaddmeasurements(handles, ParentName{1}, CPjoinstrings('Mean',SubObjectName,Fieldname), MeanVals);
            end
        else
            CPwarndlg('The Relate module is attempting to take the mean of a measurement downstream.  Be advised that unless the Relate module is placed *after* all Measurement modules, some ''Mean'' measurements will not be calculated.','Relate Module warning','replace')
        end
    end
end

% Since the label matrix starts at zero, we must include this value in
% the list to produce a label matrix image with children re-labeled to
% their parents values. This does not get saved and is only for display.
if ~isempty(ParentsOfChildren)
    ParentsOfChildrenLM = [0;ParentsOfChildren];
    NewObjectParentLabelMatrix = ParentsOfChildrenLM(SubObjectLabelMatrix+1);
else
    NewObjectParentLabelMatrix = SubObjectLabelMatrix;
end

%%%%%%%%%%%%%%%%%%%%%%%
%%% DISPLAY RESULTS %%%
%%%%%%%%%%%%%%%%%%%%%%%
drawnow

ThisModuleFigureNumber = handles.Current.(['FigureNumberForModule',CurrentModule]);

ColoredParentLabelMatrixImage = CPlabel2rgb(handles,ParentObjectLabelMatrix);
ColoredSubObjectLabelMatrixImage = CPlabel2rgb(handles,SubObjectLabelMatrix);
ColoredNewObjectParentLabelMatrix = CPlabel2rgb(handles,NewObjectParentLabelMatrix);

if any(findobj == ThisModuleFigureNumber)
    %%% Activates the appropriate figure window.
    fig_h = CPfigure(handles,'Image',ThisModuleFigureNumber);

    
    % Default image
    CPimagesc(ColoredNewObjectParentLabelMatrix,handles,ThisModuleFigureNumber);
    title('New Sub Objects')
    
    % Construct struct which holds images and figure titles
    ud(1).img = ColoredNewObjectParentLabelMatrix;
    ud(2).img = ColoredSubObjectLabelMatrixImage;
    ud(3).img = ColoredParentLabelMatrixImage;

    ud(1).title = ['New Sub Objects, cycle # ',num2str(handles.Current.SetBeingAnalyzed)];
    ud(2).title = ['Original Sub Objects, cycle # ',num2str(handles.Current.SetBeingAnalyzed)];
    ud(3).title = ['Parent Objects, cycle # ',num2str(handles.Current.SetBeingAnalyzed)];

    % Construct uicontrol text, accounting for possible StepParents
    if ~exist('StepParentObjectLabelMatrix','var')
        str = 'New Sub Objects|Original Sub Objects|Parent Objects';
    else
        str = 'New Sub Objects|Original Sub Objects|Parent Objects|StepParent Objects';
        ColoredStepParentObjectLabelMatrix = CPlabel2rgb(handles,StepParentObjectLabelMatrix);
        ud(4).img = ColoredStepParentObjectLabelMatrix;
        ud(4).title = ['StepParent Objects, cycle # ',num2str(handles.Current.SetBeingAnalyzed)];
    end
    
    % Uicontrol for displaying multiple images
    uicontrol(fig_h, 'Style', 'popup',...
        'String', str,...
        'UserData',ud,...
        'units','normalized',...
        'position',[.1 .95 .25 .04],...
        'backgroundcolor',[.7 .7 .9],...
        'Callback', @CP_ImagePopupmenu_Callback);

end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% SAVE DATA TO HANDLES STRUCTURE %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
drawnow

%%% The label matrix image is saved to the handles structure so it can be
%%% used by subsequent modules.
ColoredNewObjectParentLabelMatrixName = [ParentName{1} '_' SubObjectName];
handles = CPaddimages(handles,ColoredNewObjectParentLabelMatrixName,ColoredNewObjectParentLabelMatrix);

File: pyCellProfiler/cellprofiler/modules/relate.py
'''<b>Relate</b> assigns relationships; all objects (e.g. speckles) within a
parent object (e.g. nucleus) become its children
<hr>
Allows associating "children" objects with "parent" objects. This is
useful for counting the number of children associated with each parent,
and for calculating mean measurement values for all children that are
associated with each parent.

<p>An object will be considered a child even if the edge is the only part
touching a parent object. If an object is touching two parent objects,
the object will be assigned to the parent that shares the largest
number of pixels with the child.
'''

#CellProfiler is distributed under the GNU General Public License.
#See the accompanying file LICENSE for details.
#
#Developed by the Broad Institute
#Copyright 2003-2009
#
#Please see the AUTHORS file for credits.
#
#Website: http://www.cellprofiler.org

__version__ = "$Revision$"

import sys
import numpy as np
import scipy.ndimage as scind

import cellprofiler.cpmodule as cpm
import cellprofiler.measurements as cpmeas
import cellprofiler.settings as cps
from cellprofiler.cpmath.cpmorphology import fixup_scipy_ndimage_result as fix
from cellprofiler.cpmath.cpmorphology import centers_of_labels
from cellprofiler.cpmath.outline import outline

D_NONE = cps.DO_NOT_USE
D_CENTROID = "Centroid"
D_MINIMUM = "Minimum"
D_BOTH = "Both"

D_ALL = [D_NONE, D_CENTROID, D_MINIMUM, D_BOTH]

FF_PARENT = "Parent_%s"

FF_CHILDREN_COUNT = "Children_%s_Count"

FF_MEAN = 'Mean_%s_%s'

'''Distance category'''
C_DISTANCE = 'Distance'

'''Centroid distance feature'''
FEAT_CENTROID = 'Centroid'

'''Minimum distance feature'''
FEAT_MINIMUM = 'Minimum'

'''Centroid distance measurement (FF_DISTANCE % parent)'''
FF_CENTROID = '%s_%s_%%s' % (C_DISTANCE, FEAT_CENTROID)

'''Minimum distance measurement (FF_MINIMUM % parent)'''
FF_MINIMUM = '%s_%s_%%s' % (C_DISTANCE, FEAT_MINIMUM)

class Relate(cpm.CPModule):

    module_name = 'Relate'
    category = "Object Processing"
    variable_revision_number = 1

    def create_settings(self):
        self.sub_object_name = cps.ObjectNameSubscriber('Select the input child objects',
                                                        'None',doc="""
            The child objects are defined as those objects contained within the
            parent object.For example, to <b>Relate</b> a speckle to a containing
            nucleus, the child is the speckle object(s).""")
        self.parent_name = cps.ObjectNameSubscriber('Select the input parent objects',
                                                    'None',doc="""
            The parent objects are defined as those objects which encompass the 
            child object. For example, to <b>Relate</b> a speckle to a 
            containing nucleus, the parent is the nucleus object.""")
        self.find_parent_child_distances = cps.Choice(
            "Find distances?",
            D_ALL,doc="""
            Do you want to find the minimum distances of each child to its 
            parent?
            <br>
            <ul><li>The <i>minimum distance</i> is the distance from the 
            centroid of the child object to the closest perimeter point on
            the parent object.</li>
            <li>The <i>centroid distance</i> is the distance from the
            centroid of the child object to the centroid of the parent.
            </li></ul>""")
        self.step_parent_name = cps.ObjectNameSubscriber("Select additional objects to find distances to:", None,doc = """
                                                         You can find distances to additional objects,
                                                         or "step-parents""")
        self.wants_per_parent_means = cps.Binary('Calculate per-parent means for all child measurements?',
                                                 False,doc="""
            For every measurement that has been made of
            the children objects upstream in the pipeline, this module calculates the
            mean value of that measurement over all children and stores it as a
            measurement for the parent, as "Mean_&lt;child&gt;_&lt;category&gt;_&lt;feature&gt;". 
            For this reason, this module should be placed <i>after</i> all <b>Measure</b>
            modules that make measurements of the children objects.""")

    def settings(self):
        return [self.sub_object_name, self.parent_name, 
                self.find_parent_child_distances, self.step_parent_name,
                self.wants_per_parent_means]


    def visible_settings(self):
        # Currently, we don't support measuring distances, so those questions
        # are not shown.
        return [self.sub_object_name, self.parent_name,
                self.wants_per_parent_means]

    def run(self, workspace):
        parents = workspace.object_set.get_objects(self.parent_name.value)
        children = workspace.object_set.get_objects(self.sub_object_name.value)
        child_count, parents_of = parents.relate_children(children)
        m = workspace.measurements
        if self.wants_per_parent_means.value:
            parent_indexes = np.arange(np.max(parents.segmented))+1
            for feature_name in m.get_feature_names(self.sub_object_name.value):
                data = m.get_current_measurement(self.sub_object_name.value,
                                                 feature_name)
                if data is not None:
                    if len(parents_of) > 0:
                        means = fix(scind.mean(data.astype(float), 
                                               parents_of, parent_indexes))
                    else:
                        means = np.zeros((0,))
                    mean_feature_name = FF_MEAN%(self.sub_object_name.value,
                                                 feature_name)
                    m.add_measurement(self.parent_name.value, mean_feature_name,
                                      means)
        m.add_measurement(self.sub_object_name.value,
                          FF_PARENT%(self.parent_name.value),
                          parents_of)
        m.add_measurement(self.parent_name.value,
                          FF_CHILDREN_COUNT%(self.sub_object_name.value),
                          child_count)
        if self.find_parent_child_distances in (D_BOTH, D_CENTROID):
            self.calculate_centroid_distances(workspace)
        if self.find_parent_child_distances in (D_BOTH, D_MINIMUM):
            self.calculate_minimum_distances(workspace)
            
        if workspace.frame is not None:
            figure = workspace.create_or_find_figure(subplots=(2,2))
            figure.subplot_imshow_labels(0,0,parents.segmented,
                                         title = self.parent_name.value)
            figure.subplot_imshow_labels(1,0,children.segmented,
                                         title = self.sub_object_name.value)
            parent_labeled_children = np.zeros(children.segmented.shape, int)
            parent_labeled_children[children.segmented > 0] = \
                parents_of[children.segmented[children.segmented > 0]-1]
            figure.subplot_imshow_labels(0,1,parent_labeled_children,
                                         "%s labeled by %s"%
                                         (self.sub_object_name.value,
                                          self.parent_name.value))
    
    def calculate_centroid_distances(self, workspace):
        '''Calculate the centroid-centroid distance between parent & child'''
        meas = workspace.measurements
        assert isinstance(meas,cpmeas.Measurements)
        parent_name = self.parent_name.value
        sub_object_name = self.sub_object_name.value
        parents = workspace.object_set.get_objects(parent_name)
        children = workspace.object_set.get_objects(sub_object_name)
        parents_of = meas.get_current_measurement(sub_object_name,
                                                  FF_PARENT%(parent_name))
        pcenters = centers_of_labels(parents.segmented).transpose()
        ccenters = centers_of_labels(children.segmented).transpose()
        if pcenters.shape[0] == 0 or ccenters.shape[0] == 0:
            dist = np.array([np.NaN] * len(parents_of))
        else:
            #
            # Make indexing of parents_of be same as pcenters
            #
            parents_of = parents_of - 1
            mask = (parents_of != -1) | (parents_of > pcenters.shape[0])
            dist = np.array([np.NaN] * ccenters.shape[0])
            dist[mask] = np.sqrt(np.sum((ccenters[mask,:] - 
                                         pcenters[parents_of[mask],:])**2,1))
        meas.add_measurement(sub_object_name, FF_CENTROID % parent_name, dist)

    def calculate_minimum_distances(self, workspace):
        '''Calculate the distance from child center to parent perimeter'''
        meas = workspace.measurements
        assert isinstance(meas,cpmeas.Measurements)
        parent_name = self.parent_name.value
        sub_object_name = self.sub_object_name.value
        parents = workspace.object_set.get_objects(parent_name)
        children = workspace.object_set.get_objects(sub_object_name)
        parents_of = meas.get_current_measurement(sub_object_name,
                                                  FF_PARENT%(parent_name))
        if len(parents_of) == 0:
            dist = np.zeros((0,))
        elif np.all(parents_of == 0):
            dist = np.array([np.NaN] * len(parents_of))
        else:
            mask = parents_of > 0
            ccenters = centers_of_labels(children.segmented).transpose()
            ccenters = ccenters[mask,:]
            parents_of_masked = parents_of[mask] - 1
            pperim = outline(parents.segmented)
            #
            # Get a list of all points on the perimeter
            #
            perim_loc = np.argwhere(pperim != 0)
            #
            # Get the label # for each point
            #
            perim_idx = pperim[perim_loc[:,0],perim_loc[:,1]]
            #
            # Sort the points by label #
            #
            idx = np.lexsort((perim_loc[:,1],perim_loc[:,0],perim_idx))
            perim_loc = perim_loc[idx,:]
            perim_idx = perim_idx[idx]
            #
            # Get counts and indexes to each run of perimeter points
            #
            counts = fix(scind.sum(np.ones(len(perim_idx)),perim_idx,
                                   np.arange(1,perim_idx[-1]+1))).astype(int)
            indexes = np.cumsum(counts) - counts
            #
            # For the children, get the index and count of the parent
            #
            ccounts = counts[parents_of_masked]
            cindexes = indexes[parents_of_masked]
            #
            # Now make an array that has an element for each of that child's
            # perimeter points
            #
            clabel = np.zeros(np.sum(ccounts), int)
            #
            # cfirst is the eventual first index of each child in the
            # clabel array
            #
            cfirst = np.cumsum(ccounts) - ccounts
            clabel[cfirst[1:]] += 1
            clabel = np.cumsum(clabel)
            #
            # Make an index that runs from 0 to ccounts for each
            # child label.
            #
            cp_index = np.arange(len(clabel)) - cfirst[clabel]
            #
            # then add cindexes to get an index to the perimeter point
            #
            cp_index += cindexes[clabel]
            #
            # Now, calculate the distance from the centroid of each label
            # to each perimeter point in the parent.
            #
            dist = np.sqrt(np.sum((perim_loc[cp_index,:] - 
                                   ccenters[clabel,:])**2,1))
            #
            # Finally, find the minimum distance per child
            #
            min_dist = fix(scind.minimum(dist, clabel, np.arange(len(ccounts))))
            #
            # Account for unparented children
            #
            dist = np.array([np.NaN] * len(mask))
            dist[mask] = min_dist
        meas.add_measurement(sub_object_name, FF_MINIMUM % parent_name, dist)
            
    def validate_module(self, pipeline):
        '''Validate the module's settings
        
        Relate will complain if the children and parents are related
        by a prior module'''
        for module in pipeline.modules():
            if module == self:
                break
            parent_features = module.get_measurements(
                pipeline, self.sub_object_name.value, "Parent")
            if self.parent_name.value in (parent_features):
                raise cps.ValidationError(
                    "%s and %s were related by the %s module"%
                    (self.sub_object_name.value, self.parent_name.value,
                     module.module_name),self.parent_name)
        
    def get_measurement_columns(self, pipeline):
        '''Return the column definitions for this module's measurements'''
        columns = [(self.sub_object_name.value,
                    FF_PARENT%(self.parent_name.value),
                    cpmeas.COLTYPE_INTEGER),
                   (self.parent_name.value,
                    FF_CHILDREN_COUNT%self.sub_object_name.value,
                    cpmeas.COLTYPE_INTEGER)]
        if self.wants_per_parent_means.value:
            child_columns = pipeline.get_measurement_columns(self)
            columns += [(self.parent_name.value,
                         FF_MEAN%(self.sub_object_name.value, column[1]),
                         cpmeas.COLTYPE_FLOAT)
                        for column in child_columns
                        if column[0] == self.sub_object_name.value]
        if self.find_parent_child_distances in (D_BOTH, D_CENTROID):
            columns += [(self.sub_object_name.value,
                         FF_CENTROID % self.parent_name.value,
                         cpmeas.COLTYPE_INTEGER)]
        if self.find_parent_child_distances in (D_BOTH, D_MINIMUM):
            columns += [(self.sub_object_name.value,
                         FF_MINIMUM % self.parent_name.value,
                         cpmeas.COLTYPE_INTEGER)]
        return columns

    def get_categories(self, pipeline, object_name):
        if object_name == self.parent_name.value:
            return ["Mean_%s"%self.sub_object_name.value,
                    "Children"]
        elif object_name == self.sub_object_name.value:
            result = ["Parent"]
            if self.find_parent_child_distances != D_NONE:
                result += [C_DISTANCE]
            return result
        return []

    def get_measurements(self, pipeline, object_name, category):
        if object_name == self.parent_name.value:
            if category == "Mean_%s"%self.sub_object_name.value:
                measurements = []
                for module in pipeline.modules():
                    c = module.get_categories(self.sub_object_name.value)
                    for category in c:
                        m = module.get_measurements(self.sub_object_name.value,
                                                    category)
                        measurements += ["%s_%s"%(c,x) for x in m]
                return measurements
            elif category == "Children":
                return ["%s_Count"%self.sub_object_name.value]
        elif object_name == self.sub_object_name.value and category == "Parent":
            return [ self.parent_name.value ]
        elif (object_name == self.sub_object_name.value and 
              category == C_DISTANCE):
            result = []
            if self.find_parent_child_distances in (D_BOTH, D_CENTROID):
                result += [FF_CENTROID % self.parent_name.value]
            if self.find_parent_child_distances in (D_BOTH, D_MINIMUM):
                result += [FF_MINIMUM % self.parent_name.value]
            return result
        return []
    
    def upgrade_settings(self, setting_values, variable_revision_number, module_name, from_matlab):
        if from_matlab and variable_revision_number == 2:
            setting_values = [setting_values[0],
                              setting_values[1],
                              setting_values[2],
                              cps.YES,
                              cps.YES]
            variable_revision_number = 3
            
        if from_matlab and variable_revision_number == 3:
            setting_values = list(setting_values)
            setting_values[2] = (D_MINIMUM if setting_values[2] == cps.YES 
                                 else D_NONE)
            variable_revision_number = 4
                
        if from_matlab and variable_revision_number == 4:
            if setting_values[2] in (D_CENTROID, D_BOTH):
                sys.stderr.write("Warning: the Relate module doesn't currently support the centroid distance measurement\n")
            from_matlab = False
            variable_revision_number = 1
        return setting_values, variable_revision_number, from_matlab

